ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/Handlers.hs

```diff
@@ -144,3 +146,49 @@ handlerRequestNext currentIntersection blockTree points =
     TipPoint tip' = tip points
 
     trace = tell . pure
+
+handlerBlockFetch ::
+  forall m .
+  IOLike m =>
+  BlockTree TestBlock ->
+  ChainRange (Point TestBlock) ->
+  AdvertisedPoints ->
+  STM m (Maybe BlockFetch, [String])
+handlerBlockFetch blockTree (ChainRange from to) points =
+  runWriterT $ case points of
+    AdvertisedPoints {header = HeaderPoint (At h), block = BlockPoint (At b)} -> do
+      case BT.findFragment (headerPoint h) blockTree of
+        Just f  -> withHpFragment (f, b)
+        Nothing -> error "header point is not in the block tree"
+    AdvertisedPoints {} -> pure Nothing
+  where
+
+    withHpFragment (hpChain, bp) =
+      -- First, check whether the block point is on the same chain, and before or equal to, the header point.
+      case AF.rollback (blockPoint bp) hpChain of
+        Nothing ->
+          -- REVIEW: Should this be fatal?
+          error "Block point isn't on the same chain as header point"
+          -- pure (Just NoBlocks)
+        Just bpChain ->
+          -- Next, check whether the requested range is contained in the fragment before the block point.
+          -- REVIEW: this is Nothing if only _part_ of the range is on the current chain.
+          -- Is it correct that we don't want to send anything in this case?
+          case AF.sliceRange bpChain from to of
+            Nothing    -> do
+              trace ("Waiting for next tick for range: " ++ tersePoint from ++ " -> " ++ tersePoint to)
+              -- Finally, if we cannot serve blocks, decide whether to send NoBlocks before yielding control
+              -- to the scheduler.
+              -- If the @to@ point is not part of the chain up to HP, we must have switched to a fork, so we
+              -- need to give the client a chance to adapt the range before trying again with the next tick's
+              -- chain.
+              -- Otherwise, we simply have to wait for BP to advance sufficiently, and we block without sending
+              -- a message, to simulate a slow response.
+              case AF.withinFragmentBounds to hpChain of
+                False -> pure (Just NoBlocks)
+                True  -> pure Nothing
+            Just slice -> do
+              trace ("Sending slice " ++ terseFrag slice)
+              pure (Just (StartBatch (AF.toOldestFirst slice)))
```

Comment:
Considering how tiny and rather trivial this is, I would maybe rather write it before the `Nothing` case. Like “if we have all the blocks, we serve them; this is now out of the way; the other case is more interesting and requires more thinking.”